sql: Schema change manager approach to alter primary key

rohany opened the pull request:
The overall strategy is to take in a set of columns that will be the new primary key. Given those columns, we construct a new primary key, collect and create new indexes for all indexes that are incompatible with the new primary key, and then perform a 'swap', where we swap out the primary keys, and old version indexes with the new versions.

Design decisions:

* Add a bit to the index descriptor to mark a secondary index as "covering". This is meant to be only used in the process of primary key changes. This bit tells the encoding layer to treat this index as a primary key for all k/v encoding/decoding purposes. This allows us to get around various potential issues with decoding primary keys as secondary indexes and vice versa.
* Add a new mutation type which represents the "swap" operation, which atomically swaps fields in the table descriptor.
* There are a host of index compatibility issues that have to be worked around. 1. The new indexes being created are not compatible with the old primary key, and cannot be read from until the primary key is changed. 2. The existing indexes that are being rewritten cannot be visible once the primary key swap happens, as they will not be compatible with the new primary key.
* To get around those, the new indexes becoming readable, the primary key swap, and the old indexes becoming unreadable must all occur *in the same transaction*. The new swap mutation attempts to do that by moving the new indexes into the table descriptors indexes array from the mutation list in the same transaction that the old indexes are removed from the indexes array and placed into the mutations list.

Release note: None

rohany committed:
sql: Schema change manager approach to alter primary key

The overall strategy is to take in a set of columns that will be the new
primary key. Given those columns, we construct a new primary key, collect and
create new indexes for all indexes that are incompatible with the new primary
key, and then perform a 'swap', where we swap out the primary keys, and old
version indexes with the new versions.

Design decisions:

* Add an encoding type to the index descriptor to designate what kind of
  encoding (either secondary or primary) that a secondary index uses.
  For now, this is only used during the primary key change process. This
  encoding field tells the encoding layer how to read and write k/v
  entries for the index.
* Add a new mutation type which represents the "swap" operation, which
  atomically swaps fields in the table descriptor.
* There are a host of index compatibility issues that have to be worked
  around.
  1. The new indexes being created are not compatible with the old primary key,
     and cannot be read from until the primary key is changed.
  2. The existing indexes that are being rewritten cannot be visible once the
     primary key swap happens, as they will not be compatible with
     the new primary key.
* To get around those, the new indexes becoming readable,
  the primary key swap, and the old indexes becoming unreadable must all occur
  in the same transaction. The new swap mutation attempts to do that by moving
  the new indexes into the table descriptors indexes array from the
  mutation list in the same transaction that the old indexes are removed
  from the indexes array and placed into the mutations list.

Release note (sql change): Add experimental prototype for altering the primary
key of a table.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42462)
<!-- Reviewable:end -->

rohany commented:
```
root@127.0.0.1:59931/defaultdb> create table t (x int primary key, y int, z int not null, index i (y));
CREATE TABLE

Time: 3.336ms

root@127.0.0.1:59931/defaultdb> insert into t values (1, 2, 3);
INSERT 1

Time: 2.626ms

root@127.0.0.1:59931/defaultdb> alter table t alter primary key using columns (x, z);
ALTER TABLE

Time: 113.38ms

root@127.0.0.1:59931/defaultdb> show create t ;
  table_name |                     create_statement
+------------+----------------------------------------------------------+
  t          | CREATE TABLE t (
             |     x INT8 NOT NULL,
             |     y INT8 NULL,
             |     z INT8 NOT NULL,
             |     CONSTRAINT newprimarykey PRIMARY KEY (x ASC, z ASC),
             |     UNIQUE INDEX "primary" (x ASC) STORING (x, y, z),
             |     INDEX i (y ASC),
             |     FAMILY "primary" (x, y, z)
             | )
(1 row)

Time: 8.173ms

root@127.0.0.1:59931/defaultdb> explain select * from t@i;
  tree |    field    | description
+------+-------------+-------------+
       | distributed | true
       | vectorized  | false
  scan |             |
       | table       | t@i
       | spans       | ALL
(5 rows)

Time: 559µs

root@127.0.0.1:59931/defaultdb> select * from t@i; select * from t;
  x | y | z
+---+---+---+
  1 | 2 | 3
(1 row)

Time: 1.731ms

  x | y | z
+---+---+---+
  1 | 2 | 3
(1 row)

Time: 10µs

root@127.0.0.1:59931/defaultdb>
```

rohany commented:
I think I might need to actually start a job for the index drops rather than just placing mutations on the descriptor -- other code is failing because it expects to find jobs for each mutation id within the mutation list.

rohany commented:
Some preliminary testing does seem to show the behavior we want -- the new indexes don't seem to be visible until the swap and same for the old indexes until after the swap.

thoszhang reviewed (commented):
I left some high-level comments.

One overall thing I was thinking about is how to get this into the upcoming schema change/job framework. It seems like it wants to be two schema change jobs, but we also need the end of the first set of schema changes (to add the new indexes) to happen in the same transaction as the beginning of the second set of schema changes (to drop the old indexes). In theory, it should be doable to just create a new job in the same transaction as the final step of the first schema change (though the API would probably be less granular than manually creating the job).

As mentioned below, there are some questions around rollbacks and cancellation. If we split this into two jobs, what does it mean to try to cancel the second one?

> Reviewed 11 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-:-Lv8-Hvs6aZjb2e27adu:b-y45xdm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, @rohany, and @solongordon)

___
*[pkg/sql/alter_table.go, line 338 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7nEmY1qHDu7Zrvus5:-Lv7nEmY1qHDu7Zrvus6:b-uch2hx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/alter_table.go#L338)):*
> ```Go
> 			}
> 			// TODO (rohany): we could probably just call createOrUpdateSchemaChangeJob and writeSchemaChange
> 			//  once at the end instead of calling it repeatedly here.
> ```

Yeah, you can just wait until the end. In fact, these functions already get called at the end of `alterTableNode.startExec()`, right?
___
*[pkg/sql/alter_table.go, line 412 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7rBeb7rGnM1zoUvDD:-Lv7rBeb7rGnM1zoUvDE:bno1do0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/alter_table.go#L412)):*
> ```Go
> 				// AllocateIDs might change some information about our index, so fetch it again.
> 				// TODO (rohany): I'm doing this because some other code (create_index.go) says this could happen.
> 				//  I don't understand what could go wrong though...
> ```

I can't tell whether `allocateIndexIDs()` (which `allocateIDs()` calls) would cause problems of this type. It looks like it just mutates things in place, but I guess it's good to be safe.
___
*[pkg/sql/sem/builtins/builtins.go, line 3004 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7s4Lp9N1rmIE6rQZk:-Lv7s4Lp9N1rmIE6rQZl:b-z8phs8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sem/builtins/builtins.go#L3004)):*
> ```Go
> 				}
> 
> 				indexDesc, err := tableDesc.FindIndexByID(sqlbase.IndexID(indexID))
> ```

What is this for? The original `tableDesc.AllNonDropIndexes()[indexID-1]` looks suspicious, but I'm not sure what the intended change is.
___
*[pkg/sql/sqlbase/structured.go, line 1788 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7vFxb1AhSrSE_O3WY:-Lv7vFxc26f9GyIucKAn:b-p9zzg5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L1788)):*
> ```Go
> 			}
> 		case *DescriptorMutation_PrimaryKeySwap:
> 			// TODO (rohany): I'm not sure what we want to be doing here?
> ```

For consistency, you could check that the direction is `DescriptorMutation_NONE`, but you don't have to.
___
*[pkg/sql/sqlbase/structured.go, line 2946 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7wQ8RD8rawpY9SjwH:-Lv7wQ8S7tpJyv02XLG8:b-x4e2kg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L2946)):*
> ```Go
> 			//  I don't know if we have infrastructure / if it is sound to be updating secondary index storing columns
> 			//  when we add a column -- I don't think there is a process of also going and backfilling the secondary
> 			//  index entries and staging the addition of that column onto the storing slice of the descriptor.
> ```

Yeah, I think you do have to drop the old PK index in the general case.

If you wanted to, I suppose you could also queue the addition of a new index that's equivalent to the old PK at the same time as when you're dropping all the old indexes. But even in that case you still have to drop the existing old PK.
___
*[pkg/sql/sqlbase/structured.go, line 2977 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7xjB56k2et-yVVqog:-Lv7xjB56k2et-yVVqoh:bk2134m) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L2977)):*
> ```Go
> 			// TODO (rohany): can make this more efficient later.
> 			// TODO (rohany): this definitely has some interactions with foreign keys and other stuff that needs
> 			//  to be dealt with later.
> ```

I think what needs to happen is that when you create the new indexes, you also have to queue some foreign key mutations along with them if the old indexes had FKs, and dropping indexes should also involve queuing mutations for dropping FKs.

Unfortunately IIRC we don't actually do the second thing, so dropping indexes is slightly broken right now in general (and there is a comment in drop_index.go about this). That's not too bad since we're trying to drop the index anyway, but it would be bad if we neglected to do it when adding indexes.

(Another complication is that currently indexes "have" FKs in the sense that FKs depend on specific indexes, but this will probably no longer be true in the next few months. For now I think it's safe to just do whatever the existing CREATE INDEX and DROP INDEX implementations do.)
___
*[pkg/sql/sqlbase/structured.go, line 2987 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7zN3B5OrQKuof2g8d:-Lv7zN3B5OrQKuof2g8e:bwppf8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L2987)):*
> ```Go
> 				//  the indexes slice here as a group because they are part of the primary key swap mutation?
> 				//  Right now this is happening just because the index additions and the primary key swap mutation
> 				//  all have the same mutation ID.
> ```

Yeah, I think all the new indexes should have been inserted at this point because `MakeMutationsComplete` should process the queued mutations in order.
___
*[pkg/sql/sqlbase/structured.go, line 3003 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7zWazAM4baYlA1tx4:-Lv7zWazAM4baYlA1tx5:bly9135) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L3003)):*
> ```Go
> 				// TODO (rohany): we have to see if the asynchronous schema changing picks up on these changes.
> 				// TODO (rohany): it seems like other code wants there to be a job tied to the mutation ID
> 				//  for this group of mutations added.
> ```

Yeah, that's a good point. Unfortunately, I think to make this work you'll have to add a job manually or call `CreateOrUpdateSchemaChangeJob` or something here.
___
*[pkg/sql/sqlbase/structured.go, line 3141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7kHgz1SA4Wv-DACgf:-Lv7kHgz1SA4Wv-DACgg:blx85al) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L3141)):*
> ```Go
> }
> 
> // TODO (rohany): figure out how to do a rollback on this -- I'm not sure how???
> ```

This is a good point, though I don't think you have to do anything special for the rollback of this mutation in particular (since it doesn't cause anything to happen until the last step). I'm more concerned about how rollbacks will work for the primary key change process in general. I think the way this works now is that if we have to roll back while creating the new indexes, then we roll back all the changes up to that point (which is fine), whereas if we have to roll back while dropping the old indexes, it just tries to roll back dropping the indexes but not the primary key change itself, right?

That seems mostly fine, except that we can't actually roll back the dropping of the old indexes because they can't validly exist in a public state, so we'd have to disable the rollback somehow. It would make no sense to allow cancelling the DROP INDEX jobs.

Is that acceptable, or do you need to be able to roll back the entire thing even after the PK swap has already happened?


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
I'm not sure how it would fit into the schema change + jobs framework either. We can discuss this in person so I can get a better idea of how schema changes as jobs works!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-:-Lv8ZhR35a27ynUAo1TE:b-2hz3ey)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, and @solongordon)

___
*[pkg/sql/alter_table.go, line 338 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7nEmY1qHDu7Zrvus5:-Lv8XL6TEqjBD2WExWlX:b-tsjn8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/alter_table.go#L338)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Yeah, you can just wait until the end. In fact, these functions already get called at the end of `alterTableNode.startExec()`, right?
</blockquote></details>

yeah thats right -- i can just set descriptorChanged or whatever its called here. I thought originally that the different calls here changed something.
___
*[pkg/sql/alter_table.go, line 412 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7rBeb7rGnM1zoUvDD:-Lv8XTq-2lNUAm1k95Yw:bq0ncok) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/alter_table.go#L412)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I can't tell whether `allocateIndexIDs()` (which `allocateIDs()` calls) would cause problems of this type. It looks like it just mutates things in place, but I guess it's good to be safe.
</blockquote></details>

sounds good
___
*[pkg/sql/sem/builtins/builtins.go, line 3004 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7s4Lp9N1rmIE6rQZk:-Lv8XVg_9W5_0Gd2Z2Jj:bchalgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sem/builtins/builtins.go#L3004)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

What is this for? The original `tableDesc.AllNonDropIndexes()[indexID-1]` looks suspicious, but I'm not sure what the intended change is.
</blockquote></details>

i wrote that suspicious code a while ago and fixed it now :) I can pull it out into a separate PR however.
___
*[pkg/sql/sqlbase/structured.go, line 1788 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7vFxb1AhSrSE_O3WY:-Lv8Xc9K0I7NZm4aNqG_:b32wtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L1788)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

For consistency, you could check that the direction is `DescriptorMutation_NONE`, but you don't have to.
</blockquote></details>

ok
___
*[pkg/sql/sqlbase/structured.go, line 2946 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7wQ8RD8rawpY9SjwH:-Lv8XebV45ttlZvUUvh8:b-3ubi6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L2946)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Yeah, I think you do have to drop the old PK index in the general case.

If you wanted to, I suppose you could also queue the addition of a new index that's equivalent to the old PK at the same time as when you're dropping all the old indexes. But even in that case you still have to drop the existing old PK.
</blockquote></details>

Thats something that we are considering -- dropping the existing old PK shouldn't cause any problems. Additionally, we were thinking about giving the syntax some way to saying whether you want the old PK to stay around.
___
*[pkg/sql/sqlbase/structured.go, line 2977 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7xjB56k2et-yVVqog:-Lv8XqNH0q0QEy75Bbyi:bo60m75) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L2977)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think what needs to happen is that when you create the new indexes, you also have to queue some foreign key mutations along with them if the old indexes had FKs, and dropping indexes should also involve queuing mutations for dropping FKs.

Unfortunately IIRC we don't actually do the second thing, so dropping indexes is slightly broken right now in general (and there is a comment in drop_index.go about this). That's not too bad since we're trying to drop the index anyway, but it would be bad if we neglected to do it when adding indexes.

(Another complication is that currently indexes "have" FKs in the sense that FKs depend on specific indexes, but this will probably no longer be true in the next few months. For now I think it's safe to just do whatever the existing CREATE INDEX and DROP INDEX implementations do.)
</blockquote></details>

I talked with @jordanlewis about this, and we came to the conclusion that we wouldn't need to do anything here, because the new indexes index the same things as the old indexes, so the paths that choose which index to use for foreign key operations will be satisfied. The tricky part here would be if a FK depends on the old primary index, in which case we can make a new index to still satisfy the fk.
___
*[pkg/sql/sqlbase/structured.go, line 2987 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7zN3B5OrQKuof2g8d:-Lv8YEyU0FGENRY1WWtU:bks2r22) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L2987)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Yeah, I think all the new indexes should have been inserted at this point because `MakeMutationsComplete` should process the queued mutations in order.
</blockquote></details>

ok -- I can leave a more detailed comment about why this is the case.
___
*[pkg/sql/sqlbase/structured.go, line 3003 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7zWazAM4baYlA1tx4:-Lv8YL153qN1QZD7fIs4:b-iue7pw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L3003)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Yeah, that's a good point. Unfortunately, I think to make this work you'll have to add a job manually or call `CreateOrUpdateSchemaChangeJob` or something here.
</blockquote></details>

ok, I can look into that.
___
*[pkg/sql/sqlbase/structured.go, line 3141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7kHgz1SA4Wv-DACgf:-Lv8ZV220ANYVmVKsEcx:bhoell6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L3141)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

This is a good point, though I don't think you have to do anything special for the rollback of this mutation in particular (since it doesn't cause anything to happen until the last step). I'm more concerned about how rollbacks will work for the primary key change process in general. I think the way this works now is that if we have to roll back while creating the new indexes, then we roll back all the changes up to that point (which is fine), whereas if we have to roll back while dropping the old indexes, it just tries to roll back dropping the indexes but not the primary key change itself, right?

That seems mostly fine, except that we can't actually roll back the dropping of the old indexes because they can't validly exist in a public state, so we'd have to disable the rollback somehow. It would make no sense to allow cancelling the DROP INDEX jobs.

Is that acceptable, or do you need to be able to roll back the entire thing even after the PK swap has already happened?
</blockquote></details>

I don't really have any idea how rollbacks work, or what we would want/expect from the primary key change process with rollbacks -- we can discuss this in person.


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, and @solongordon)

___
*[pkg/sql/alter_table.go, line 338 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7nEmY1qHDu7Zrvus5:-LvBNS1pCFhzF3OSUSjR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/alter_table.go#L338)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

yeah thats right -- i can just set descriptorChanged or whatever its called here. I thought originally that the different calls here changed something.
</blockquote></details>

Done.
___
*[pkg/sql/alter_table.go, line 412 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7rBeb7rGnM1zoUvDD:-LvBNlaICrz36DYhFUcD:b-4d74wp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/alter_table.go#L412)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

sounds good
</blockquote></details>

I looked into this more, and I think i was confused the first time i wrote this code. The comment in create_index.go is 
```
	// The index name may have changed as a result of
	// AllocateIDs(). Retrieve it for the event log below.
```
which is true, because an unnamed index gets allocated a name. However, that is all that changes here.
___
*[pkg/sql/sem/builtins/builtins.go, line 3004 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7s4Lp9N1rmIE6rQZk:-LvBO_E_8eXglZazDuZo:b-38asw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sem/builtins/builtins.go#L3004)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

i wrote that suspicious code a while ago and fixed it now :) I can pull it out into a separate PR however.
</blockquote></details>

Turns out I already landed code that fixed this. 
___
*[pkg/sql/sqlbase/structured.go, line 3003 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7zWazAM4baYlA1tx4:-LvBS2ZW9jnmFPlmx_--:b-ut63cc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L3003)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

ok, I can look into that.
</blockquote></details>

I think I'll have to manually make a job here -- I don't think I can plumb a planner into this method.


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, and @solongordon)

___
*[pkg/sql/sqlbase/structured.go, line 3003 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7zWazAM4baYlA1tx4:-LvBzPcREgKYo56fcL-o:b469gq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L3003)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I think I'll have to manually make a job here -- I don't think I can plumb a planner into this method.
</blockquote></details>

Ack i'm not sure I can even get access to alot of the things that the planner uses in CreateOrUpdateSchemaChangeJob. Could remembering the mutation ID of the initial group of index additions and using that on the index drops work?


<!-- Sent from Reviewable.io -->

rohany commented:
@lucy-zhang I think this is ready for a close review -> merging as a prototype. I've added a bunch of different schema changer style tests, and added checks/bail outs for the things that we currently don't support.

thoszhang reviewed (commented):
> Reviewed 16 of 19 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, @rohany, and @solongordon)

___
*[pkg/sql/alter_table.go, line 368 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lvkr0t8DX6oIztjNt2B:-Lvkr0t8DX6oIztjNt2C:bc47o8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/82a7f6da7e765ca449d5ee0649bfa21a78e045ce/pkg/sql/alter_table.go#L368)):*
> ```Go
> 			// TODO (rohany,solongordon): This should be changed to see if any of the InboundFK's depend on indexes
> 			//  that are being rewritten.
> 			if len(n.tableDesc.InboundFKs) > 0 {
> ```

Why doesn't this problem also exist with the outbound FKs? I thought that for outbound FKs that depend on columns being rewritten, we would also have to update the index ID on the referenced side.
___
*[pkg/sql/alter_table.go, line 379 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwHUhr1cykdiUsMAgq:-LvwHUhr1cykdiUsMAgr:b5gbu78) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/alter_table.go#L379)):*
> ```Go
> 			if len(n.tableDesc.PrimaryIndex.ColumnNames) == 1 && n.tableDesc.PrimaryIndex.ColumnNames[0] != "rowid" {
> 				oldPrimaryIndexCopy := protoutil.Clone(&n.tableDesc.PrimaryIndex).(*sqlbase.IndexDescriptor)
> 				// TODO (rohany): what if this name clashes with an existing index name?
> ```

I think this should get updated before the PR gets merged so that we can test with multiple sequential PK changes on the same table right away, but it doesn't really matter what, as long as we do something reasonable. We could just add a random string at the end.
___
*[pkg/sql/alter_table.go, line 904 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwHN1I05uuf-ZgjjOL:-LvwHN1I05uuf-ZgjjOM:bn9jlfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/alter_table.go#L904)):*
> ```Go
> // addIndexWithSpecificPrimaryKey adds an index mutation into the given table descriptor, but sets up
> // the index with ExtraColumnIDs from the given index, rather than the table's primary key.
> func addIndexWithSpecificPrimaryKey(
> ```

I think this should have `Mutation` in the name to match `AddIndexMutation`.
___
*[pkg/sql/schema_changer.go, line 1319 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwIGLc0rsy_eDjv0rS:-LvwIGLc0rsy_eDjv0rT:bh7q5rg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1319)):*
> ```Go
> 				// a job for the index deletion mutations that the primary key swap mutation added, if any.
> 				mutationID := scDesc.ClusterVersion.NextMutationID
> 				// TODO (rohany): figure out what the deal with the ResumeSpanList is.
> ```

The `ResumeSpanList` field is how the backfiller tracks its progress, so setting it to the primary key span is fine. (I don't even know if it matters when dropping indexes.)
___
*[pkg/sql/schema_changer.go, line 1333 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwJ9OS1HKRrVsqS0MG:-LvwJ9OS1HKRrVsqS0MH:b4o40hl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1333)):*
> ```Go
> 					Description: "Primary key swap index cleanup job",
> 					// TODO (rohany): what user should I put here?
> 					Username:      "????",
> ```

Can you get the username from the previous job's payload?

I don't really know what the job's username is used for. We also use `security.RootUser` for some of the bulk IO/schema change jobs. I think @dt would know.
___
*[pkg/sql/schema_changer.go, line 1570 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwM6f39Jn_LJSdaNTp:-LvwM6f39Jn_LJSdaNTq:bro3rh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1570)):*
> ```Go
> 		// In the case that there are any primary key change mutations with mutationID
> 		// equal to sc.mutationID, we need to remove those mutations from the mutations
> 		// list, since they don't have any cleanup to perform.
> ```

Why do we need to remove the PK change mutations explicitly, if there's no action needed on rollback?

Also, can there ever be multiple PK change mutations with the same mutation ID? That would seem to imply they were started in the same transaction, which seems like something we'd want to disallow.
___
*[pkg/sql/schema_changer.go, line 1581 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwMGxg8vXEka1rYGpc:-LvwMGxg8vXEka1rYGpd:b-k780tb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1581)):*
> ```Go
> 		}
> 		for _, del := range primaryKeySwapMutations {
> 			scDesc.Mutations = append(scDesc.Mutations[:del], scDesc.Mutations[del+1:]...)
> ```

Does this loop work if you're removing elements at multiple indexes?
___
*[pkg/sql/schema_changer_test.go, line 2408 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwO5GuBV_ylaZ0uIQ4:-LvwO5GvBigfBu_CH5Pl:b-uri9bz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer_test.go#L2408)):*
> ```Go
> 		// conflict handling. Since we are changing the primary key of the table to
> 		// be using k as the primary key, we disable the UPSERT statements in the test.
> 		false,
> ```

I'm not following, sorry. What is the error, exactly? Does it have to do with the unique constraint on `k`?
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwOyxd2RkEAuz2v8q3:-LvwOyxd2RkEAuz2v8q4:b-7bcvh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L1)):*
> ```
> 
> ```

nit: extra newline
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwOc5V-QWNFVztgJVs:-LvwOc5V-QWNFVztgJVt:bgbvuhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L20)):*
> ```
> # Test that composite values are encoded correctly in covering indexes.
> statement ok
> CREATE TABLE tComposite (x INT PRIMARY KEY, y DECIMAL NOT NULL, FAMILY (x, y));
> ```

nit: I think we usually use snakecase table names in logic tests, like `t_composite`.
___
*[pkg/sql/sqlbase/structured.go, line 3003 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7zWazAM4baYlA1tx4:-LvwTarq8CANz7xCuG5a:bu4sdps) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L3003)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Ack i'm not sure I can even get access to alot of the things that the planner uses in CreateOrUpdateSchemaChangeJob. Could remembering the mutation ID of the initial group of index additions and using that on the index drops work?
</blockquote></details>

The job is now being created in schema_changer.go, right?


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, @rohany, and @solongordon)

___
*[pkg/sql/alter_table.go, line 368 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lvkr0t8DX6oIztjNt2B:-LvztpTAC3lkQhArP1bZ:b-yeh3t2) ([raw file](https://github.com/cockroachdb/cockroach/blob/82a7f6da7e765ca449d5ee0649bfa21a78e045ce/pkg/sql/alter_table.go#L368)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Why doesn't this problem also exist with the outbound FKs? I thought that for outbound FKs that depend on columns being rewritten, we would also have to update the index ID on the referenced side.
</blockquote></details>

If i'm understanding correctly, outbound FK's are pointing from this table onto another table, which means that its an index in another table that is important to the FK. The other table isn't changing, so we don't need to do anything in this case?
___
*[pkg/sql/schema_changer.go, line 1570 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwM6f39Jn_LJSdaNTp:-LvzvzhxA9PYUcNsD13d:b-ti8utm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1570)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Why do we need to remove the PK change mutations explicitly, if there's no action needed on rollback?

Also, can there ever be multiple PK change mutations with the same mutation ID? That would seem to imply they were started in the same transaction, which seems like something we'd want to disallow.
</blockquote></details>

I might be misunderstanding rollbacks -- it seemed to me that when a rollback occurred, all the mutations were reversed, and then the schema changer proceeded. In that case, we would want to remove this PK swap mutation so that it doesn't get processed right?

I don't think we should allow that -- i can add a check for this in alter_table.go
___
*[pkg/sql/schema_changer.go, line 1581 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwMGxg8vXEka1rYGpc:-LvzwCk35HJhgjxQmLMN:brqgszr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1581)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Does this loop work if you're removing elements at multiple indexes?
</blockquote></details>

why shouldn't it? the slice being looped over is different than the one that is being mutated
___
*[pkg/sql/schema_changer_test.go, line 2408 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwO5GuBV_ylaZ0uIQ4:-LvzwQ592g9u38y2ZuME:bwiuhbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer_test.go#L2408)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I'm not following, sorry. What is the error, exactly? Does it have to do with the unique constraint on `k`?
</blockquote></details>

Here is the docs that I looked at -- https://www.cockroachlabs.com/docs/stable/upsert.html#upsert-that-fails-conflict-on-non-primary-key. The operations that the test ran use upsert, but assuming that `k` is the primary key, so conflicts on k can proceed with the upsert. When `k` isn't the primary key, these conflicts lead to errors, which makes sense to me.
___
*[pkg/sql/sqlbase/structured.go, line 3003 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7zWazAM4baYlA1tx4:-Lvzwv2oCgMFbLthydLd:bv3nzm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L3003)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

The job is now being created in schema_changer.go, right?
</blockquote></details>

yup, this is outdated.


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, and @solongordon)

___
*[pkg/sql/alter_table.go, line 379 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwHUhr1cykdiUsMAgq:-Lw0drfH8whxYn36F1ft:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/alter_table.go#L379)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think this should get updated before the PR gets merged so that we can test with multiple sequential PK changes on the same table right away, but it doesn't really matter what, as long as we do something reasonable. We could just add a random string at the end.
</blockquote></details>

Done.
___
*[pkg/sql/alter_table.go, line 904 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwHN1I05uuf-ZgjjOL:-Lw0XYWN-zmny_GhTY6v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/alter_table.go#L904)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think this should have `Mutation` in the name to match `AddIndexMutation`.
</blockquote></details>

Done.
___
*[pkg/sql/schema_changer.go, line 1319 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwIGLc0rsy_eDjv0rS:-Lw0XfvG91ML9XougOf_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1319)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

The `ResumeSpanList` field is how the backfiller tracks its progress, so setting it to the primary key span is fine. (I don't even know if it matters when dropping indexes.)
</blockquote></details>

Done.
___
*[pkg/sql/schema_changer.go, line 1333 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwJ9OS1HKRrVsqS0MG:-Lw0ZA9S3trjcawIj_Wd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1333)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Can you get the username from the previous job's payload?

I don't really know what the job's username is used for. We also use `security.RootUser` for some of the bulk IO/schema change jobs. I think @dt would know.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, and @solongordon)

___
*[pkg/sql/schema_changer_test.go, line 2515 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lw1yHex6uijEfFfAWrS:-Lw1yHex6uijEfFfAWrT:btd90fj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1858be681ced1e3b6bd87adceb5e2cdfd26d9b84/pkg/sql/schema_changer_test.go#L2515)):*
> ```Go
> 	// Ensure that the writes from the partial new indexes are cleaned up.
> 	testutils.SucceedsSoon(t, func() error {
> 		return checkTableKeyCount(ctx, kvDB, 1, maxValue)
> ```

This test seems to be kinda flaky on CI -- any ideas on how to deflake it?


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, and @solongordon)

___
*[pkg/sql/schema_changer_test.go, line 2515 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lw1yHex6uijEfFfAWrS:-LwFKav_8eoj80r0HD6p:b2z75xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1858be681ced1e3b6bd87adceb5e2cdfd26d9b84/pkg/sql/schema_changer_test.go#L2515)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

This test seems to be kinda flaky on CI -- any ideas on how to deflake it?
</blockquote></details>

I think this is fixed -- i added a GC config for the table.


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, and @solongordon)

___
*[pkg/sql/alter_table.go, line 368 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lvkr0t8DX6oIztjNt2B:-LwFtzHB1xD66xW13G6A:b-n1o79y) ([raw file](https://github.com/cockroachdb/cockroach/blob/82a7f6da7e765ca449d5ee0649bfa21a78e045ce/pkg/sql/alter_table.go#L368)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

If i'm understanding correctly, outbound FK's are pointing from this table onto another table, which means that its an index in another table that is important to the FK. The other table isn't changing, so we don't need to do anything in this case?
</blockquote></details>

You're right, I had some misunderstanding of what was going on. updated!


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> Reviewed 1 of 20 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, @rohany, and @solongordon)

___
*[pkg/sql/alter_table.go, line 324 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LwUeLh2Azrp0Mp-8GrS:-LwUeLh2Azrp0Mp-8GrT:b-6yn0j7) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f6fdf974dfbeddeb1c239ccce702a01443fad3/pkg/sql/alter_table.go#L324)):*
> ```Go
> 				if desc := n.tableDesc.Mutations[i].GetPrimaryKeySwap(); desc != nil &&
> 					n.tableDesc.Mutations[i].MutationID == currentMutationID {
> 					// TODO (rohany): what's the best pgcode for this?
> ```

Good question. Maybe @knz knows
___
*[pkg/sql/alter_table.go, line 352 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LwUhc3k7CF0I6bVC9vJ:-LwUhc3l42V6dtlknBah:bk3lv5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f6fdf974dfbeddeb1c239ccce702a01443fad3/pkg/sql/alter_table.go#L352)):*
> ```Go
> 			name := "newprimarykey"
> 			for try := 1; nameExists(name); try++ {
> 				name = fmt.Sprintf("%s#%d", name, try)
> ```

Does this make names of the form `newprimarykey`, `newprimarykey#1`, `newprimarykey#1#2`, ...?  Maybe I'm missing something here.

Also, I think for consistency with other generated names we'll want `new_primary_key_1`, etc.
___
*[pkg/sql/alter_table.go, line 404 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LwUlOu808I4QrHfibYL:-LwUlOu808I4QrHfibYM:benpsw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f6fdf974dfbeddeb1c239ccce702a01443fad3/pkg/sql/alter_table.go#L404)):*
> ```Go
> 				name := "old_primary_key"
> 				for try := 1; nameExists(name); try++ {
> 					name = fmt.Sprintf("%s#%d", name, try)
> ```

Same comment as above.
___
*[pkg/sql/schema_changer.go, line 1319 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwIGLc0rsy_eDjv0rS:-LwUlvl8BFV3-gHq15LX:bpmyny7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1319)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Hmm, a question just occurred to me. _Which_ primary key span is this?
___
*[pkg/sql/schema_changer.go, line 1570 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwM6f39Jn_LJSdaNTp:-LwUnb4MDs-Y_U0pf7nJ:b-var5al) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1570)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I might be misunderstanding rollbacks -- it seemed to me that when a rollback occurred, all the mutations were reversed, and then the schema changer proceeded. In that case, we would want to remove this PK swap mutation so that it doesn't get processed right?

I don't think we should allow that -- i can add a check for this in alter_table.go
</blockquote></details>

Hmm, I think you're right. I'd been thinking in terms of the "add" mutation doing something and the "drop" mutation doing nothing, but I guess there's no concept of a "direction" so it works either way. This doesn't seem quite natural to me. It seems to me like the 3 options are

1. do it like this
2. instead of deleting the mutation, keep it but do nothing in the backfiller if we're rolling back
3. instead of deleting the mutation, change the "none" mutation direction state to "add"/"drop" and process them differently

The advantage I see of keeping the mutation in the mutations slice is that you know the schema changes you're rolling back originally came from a PK change, and you don't lose the useful metadata on the proto. But I haven't thought that much about this.
___
*[pkg/sql/schema_changer.go, line 1581 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwMGxg8vXEka1rYGpc:-LwUnEZ016N8g1OfMoG-:b-ex13po) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1581)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

why shouldn't it? the slice being looped over is different than the one that is being mutated
</blockquote></details>

I'm confused. Doesn't the index of any given mutation in the slice change after you've already removed an element before it? In any case I guess this is no longer relevant.
___
*[pkg/sql/schema_changer_test.go, line 2536 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LwUoviq7U8doT8mGLAZ:-LwUoviq7U8doT8mGLA_:bu1s067) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f6fdf974dfbeddeb1c239ccce702a01443fad3/pkg/sql/schema_changer_test.go#L2536)):*
> ```Go
> // TestMultiplePrimaryKeyChanges ensures that we can run many primary key changes back to back.
> // We cannot run this in a logic test because we need the async schema changer to run so that
> // the following primary key change occurs quickly.
> ```

It's waiting for the indexes to be dropped (but not GCed), right? What exactly happens if you try to do this in a logic test?


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, @rohany, and @solongordon)

___
*[pkg/sql/alter_table.go, line 352 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LwUhc3k7CF0I6bVC9vJ:-LwUqQnYA0g2cj15KuSy:bhy20we) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f6fdf974dfbeddeb1c239ccce702a01443fad3/pkg/sql/alter_table.go#L352)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Does this make names of the form `newprimarykey`, `newprimarykey#1`, `newprimarykey#1#2`, ...?  Maybe I'm missing something here.

Also, I think for consistency with other generated names we'll want `new_primary_key_1`, etc.
</blockquote></details>

oh woops... you're right. I'll update that to use underscores too.
___
*[pkg/sql/alter_table.go, line 404 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LwUlOu808I4QrHfibYL:-LwUrJJ12zoZBEgS7TXW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f6fdf974dfbeddeb1c239ccce702a01443fad3/pkg/sql/alter_table.go#L404)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Same comment as above.
</blockquote></details>

Done.
___
*[pkg/sql/schema_changer.go, line 1319 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwIGLc0rsy_eDjv0rS:-LwV-BO30kTAXZ33Hvnr:b-xtzhm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1319)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Hmm, a question just occurred to me. _Which_ primary key span is this?
</blockquote></details>

This is the new primary key -- does that matter?
___
*[pkg/sql/schema_changer.go, line 1570 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwM6f39Jn_LJSdaNTp:-LwV-JuUAtQQjN38XdA0:br4nrm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1570)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Hmm, I think you're right. I'd been thinking in terms of the "add" mutation doing something and the "drop" mutation doing nothing, but I guess there's no concept of a "direction" so it works either way. This doesn't seem quite natural to me. It seems to me like the 3 options are

1. do it like this
2. instead of deleting the mutation, keep it but do nothing in the backfiller if we're rolling back
3. instead of deleting the mutation, change the "none" mutation direction state to "add"/"drop" and process them differently

The advantage I see of keeping the mutation in the mutations slice is that you know the schema changes you're rolling back originally came from a PK change, and you don't lose the useful metadata on the proto. But I haven't thought that much about this.
</blockquote></details>

I think 3 makes good sense.
___
*[pkg/sql/schema_changer.go, line 1581 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwMGxg8vXEka1rYGpc:-LwV0x66557ERSM1FGjD:b34uopr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1581)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I'm confused. Doesn't the index of any given mutation in the slice change after you've already removed an element before it? In any case I guess this is no longer relevant.
</blockquote></details>

This has been removed.
___
*[pkg/sql/schema_changer_test.go, line 2536 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LwUoviq7U8doT8mGLAZ:-LwV10qH88vlHDEUyh1L:bwzwnod) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f6fdf974dfbeddeb1c239ccce702a01443fad3/pkg/sql/schema_changer_test.go#L2536)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

It's waiting for the indexes to be dropped (but not GCed), right? What exactly happens if you try to do this in a logic test?
</blockquote></details>

It stalls until the async schema changer picks up the drop index schema change, so it seems to slow down logic tests alot. I bet this can be fixed in the new jobs framework.
___
*[pkg/sql/sqlbase/structured.go, line 3141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-Lv7kHgz1SA4Wv-DACgf:-LwV1BJS0Zvs6AjM9HrX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c174808f01ce3eefbb410668b6aca81be4735c84/pkg/sql/sqlbase/structured.go#L3141)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I don't really have any idea how rollbacks work, or what we would want/expect from the primary key change process with rollbacks -- we can discuss this in person.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, and @solongordon)

___
*[pkg/sql/alter_table.go, line 324 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LwUeLh2Azrp0Mp-8GrS:-LwVHGrCBT-AnSrWPVnd:b-o355h1) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f6fdf974dfbeddeb1c239ccce702a01443fad3/pkg/sql/alter_table.go#L324)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Good question. Maybe @knz knows
</blockquote></details>

I would go for `return unimplemented.New("multiple-alter-pk-in-txn", "cannot perform ...")`

or perhaps `unimplemented.NewWithIssue` after you file a follow-up issue and tag it with C-enhancement X-anchored-telemetry.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, and @solongordon)

___
*[pkg/sql/alter_table.go, line 324 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LwUeLh2Azrp0Mp-8GrS:-LwVHnqZ0ulCluxfvtBE:blw0xvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f6fdf974dfbeddeb1c239ccce702a01443fad3/pkg/sql/alter_table.go#L324)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I would go for `return unimplemented.New("multiple-alter-pk-in-txn", "cannot perform ...")`

or perhaps `unimplemented.NewWithIssue` after you file a follow-up issue and tag it with C-enhancement X-anchored-telemetry.
</blockquote></details>

The advantage of an open issue is that we can later fix it (support the feature) or we can explain on the issue how to workaround, with links to docs etc.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
lgtm, minor comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-:-LxwQZS27BouU3EEQQGM:b-k068r8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, @rohany, and @solongordon)

___
*[pkg/sql/alter_table.go, line 404 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LwUlOu808I4QrHfibYL:-LxwOoDTD5qtfZiUuCmM:b8vce3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f6fdf974dfbeddeb1c239ccce702a01443fad3/pkg/sql/alter_table.go#L404)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Should this also have an underscore?
___
*[pkg/sql/alter_table.go, line 325 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LxwOdeP3wHKE0M-Qag0:-LxwOdeP3wHKE0M-Qag1:b6h2s4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/9de47fbe1897dc4dcbc230e502eefcfe11a8e460/pkg/sql/alter_table.go#L325)):*
> ```Go
> 					n.tableDesc.Mutations[i].MutationID == currentMutationID {
> 					return unimplemented.NewWithIssue(
> 						43376, "multiple primary key changes in the same transaction is unsupported")
> ```

`are unsupported` maybe
___
*[pkg/sql/schema_changer.go, line 1319 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LvwIGLc0rsy_eDjv0rS:-LxwOuwE1WznNe_CBJBZ:bfj4yl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bfa4499d106bba0ea1ab0b060f6e68180f75e14/pkg/sql/schema_changer.go#L1319)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

This is the new primary key -- does that matter?
</blockquote></details>

I think that should be fine.
___
*[pkg/sql/schema_changer.go, line 1837 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LxwQ8VWBeVeOvNtSKsw:-LxwQ8VWBeVeOvNtSKsx:bkq9vze) ([raw file](https://github.com/cockroachdb/cockroach/blob/9de47fbe1897dc4dcbc230e502eefcfe11a8e460/pkg/sql/schema_changer.go#L1837)):*
> ```Go
> 		}
> 		// PrimaryKeySwap doesn't have a concept of the state machine.
> 		if pkSwap := mutation.GetPrimaryKeySwap(); pkSwap != nil {
> ```

Why does the primary key swap case need special handling here?


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> Reviewed 16 of 20 files at r5, 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @rohany, and @solongordon)



<!-- Sent from Reviewable.io -->

rohany commented:
Another, possibly cleaner solution is to just arbitrarily pick one of the families that contains primary key columns to *be the sentinel* for that family, and replace all point lookups on family 0 to use that sentinel. This sentinel is stable because primary key columns can't be dropped, and might be easy to change in the primary key change process, but I haven't worked out the details.

rohany commented:
bors r=lucy-zhang

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42462)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @lucy-zhang, @rohany, and @solongordon)

___
*[pkg/sql/alter_table.go, line 404 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42462#-LwUlOu808I4QrHfibYL:-Lxx8nuFAruMmVUQjfMW:b-do4xsy) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f6fdf974dfbeddeb1c239ccce702a01443fad3/pkg/sql/alter_table.go#L404)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Should this also have an underscore?
</blockquote></details>

I meant instead of the `#`.


<!-- Sent from Reviewable.io -->

rohany commented:
yikes i'm dumb. I'll open a PR.
